fix(rum-core): capture all spans as part of page-load transaction

vigneshshanmugam opened the pull request:
+ fix elastic/apm-agent-rum-js#240 

+ For page load transactions, we have to adjust the transaction start/end depending on the navigation & resource timing spans.Deciding it as part of transaction.end would  would exclude the navigation and resource timing spans since we create the nav and res spans later after the load event has happened and we might consider them invalid since transaction.end would be lesser than those spans. 

+ I am not sure how we missed this, But this is a huge issue if the user has any custom/external spans that is finished before the page-load event fires.

jahtalab commented on packages/rum-core/src/performance-monitoring/transaction.js:
```diff
@@ -132,8 +134,18 @@ class Transaction extends SpanBase {
+     */
+    if (this.type !== PAGE_LOAD) {
+      this._adjustStartToEarliestSpan()
+      this._adjustEndToLatestSpan()
```
We still call these two in the `capture-navigation-timing.js`, not sure why this would solve the issue.
In the capture-navigation-timing we just add the spans and then adjust the start/end of the transaction to be compatible with the spans that were added for resource and navigation.

vigneshshanmugam replied:
The main point is to do it only in https://github.com/elastic/apm-agent-rum-js/blob/c162a1252f034c2ec59d5c5ce349412857d817aa/packages/rum-core/src/performance-monitoring/capture-hard-navigation.js#L217-L218 for the page-load transaction after we have calculated the nav timing spans. 

Doing it on the `transacton.js` prevents the nav/res timing spans from getting added which is the issue that we are fixing.

vigneshshanmugam commented:
Docs build failure is unrelated.

jahtalab commented:
@vigneshshanmugam , I'm still not sure why this bug only existed on chrome 49 (sorry if I'm overlooking something here).

vigneshshanmugam commented:
Sorry if I did not explain it properly in the description/discussion, The but exists in all browsers. The number of spans that should exist in page load is 8. We we're seeing 4 in other browsers l, whereas in chrome 49 we see only 2 sometimes..Most of the ajax/nav timing spans fails to get captured bcoz of the bug which is fixed now. The spans is not still 8 because sometimes the spans like dns/load might be 0 since the due to network cache and other behaviours. So the test is reasonably adjusted to check for more than 4 instead of the previous value 2. 

If you want more details, you can print the console logs before and after this fix. Please watch the number of spans getting added to the transactions.

Let me know if you need more info.

vigneshshanmugam commented:
To give more context, Please check the below console logs for before and after the fix. 

## Tested on Chrome 74. 

### Before 

<img width="886" alt="Screenshot 2019-06-05 at 12 52 24" src="https://user-images.githubusercontent.com/3902525/58951315-198a1c00-8791-11e9-83fd-f4f5e54f4597.png">

### After

<img width="888" alt="Screenshot 2019-06-05 at 12 58 22" src="https://user-images.githubusercontent.com/3902525/58951516-a0d78f80-8791-11e9-983f-913455af2d06.png">

jahtalab commented:
Thanks @vigneshshanmugam! I was thinking about how we can refactor this part to avoid this kind of bugs! At the moment it seems we’re introducing a special case into the transaction class which until now didn't have any specific code about transaction type.

jahtalab commented:
Had a meeting with @vigneshshanmugam and we decided to refactor adjusting the transaction to it's spans into the PerformanceMonitoring class!

vigneshshanmugam commented:
@jahtalab The refactor is done and I have moved the tests around closer to the functionality of the class. Also added extra guard for the general usecase test to catch bugs.

jahtalab commented on packages/rum-core/test/performance-monitoring/transaction.spec.js:
```diff
@@ -168,31 +125,6 @@ describe('transaction.Transaction', function() {
     expect(span).toBe(undefined)
   })
 
-  it('should not produce negative durations while adjusting to the spans', function() {
```
Shouldn't we move this test to performance monitoring as well?

vigneshshanmugam replied:
Oh yeah might have missed, Let me add it

jahtalab commented on packages/rum-core/test/performance-monitoring/performance-monitoring.spec.js:
```diff
@@ -261,13 +261,68 @@ describe('PerformanceMonitoring', function() {
     })
     var span = tr.startSpan('span1', 'span1type')
     span.end()
-    span._end += 10
```
I think this was added to avoid having the span duration to be 0 since we start and end the span quickly.

vigneshshanmugam replied:
Yeah I had the same impression, It makes the span truncated since transaction end is lesser. But I can call performance.adjustTime and pass the test. We pay some time for creating the span so I thought its reasonable to remove it .WDYT?

jahtalab replied:
But `performanceMonitoring.createTransactionPayload` calls the adjustTime!

Also, it's ok to change the tr._end after `expect(tr._end).toBeDefined()` to make sure that it includes the span!

vigneshshanmugam replied:
Did the changes.
